File: frontend/src/domain/simulationValueObjects.ts
export const SIMULATION_N_SIMS_MIN = 1_000;
export const SIMULATION_N_SIMS_MAX = 200_000;
export const SIMULATION_TARGET_WEEKS_MIN = 1;
export const SIMULATION_HORIZON_WEEKS_MAX = 521;
export const SIMULATION_THROUGHPUT_SAMPLES_MIN = 6;
export const SIMULATION_THROUGHPUT_SAMPLES_MAX = 521;
export const SIMULATION_BACKLOG_SIZE_MIN = 1;
export const SIMULATION_BACKLOG_SIZE_MAX = 1_000_000;
export const SIMULATION_SEED_MIN = 0;
export const SIMULATION_SEED_MAX = 0xffffffff;

export type SimulationMode = "backlog_to_weeks" | "weeks_to_items";
export type PercentileKey = "P50" | "P70" | "P90";
export type ThroughputReliabilityLabel = "fiable" | "incertain" | "fragile" | "non fiable";

declare const simulationSeedBrand: unique symbol;
declare const simulationCountBrand: unique symbol;
declare const backlogSizeBrand: unique symbol;
declare const simulationHorizonBrand: unique symbol;
declare const throughputSamplesBrand: unique symbol;

export type SimulationSeed = number & { readonly [simulationSeedBrand]?: never };
export type SimulationCount = number & { readonly [simulationCountBrand]?: never };
export type BacklogSize = number & { readonly [backlogSizeBrand]?: never };
export type SimulationHorizon = number & { readonly [simulationHorizonBrand]?: never };

export type ThroughputSamples = Readonly<{
  rawValues: readonly number[];
  usableValues: readonly number[];
  includeZeroWeeks: boolean;
  readonly [throughputSamplesBrand]: true;
}>;

export type SimulationPercentiles = Readonly<Partial<Record<PercentileKey, number>>>;

export type HistogramBucket = Readonly<{
  x: number;
  count: number;
}>;

export type SimulationHistogram = readonly HistogramBucket[];

export type CompletionSummary = Readonly<{
  completedCount: number;
  censoredCount: number;
  censoredRate: number;
  horizonWeeks: number;
}>;

export type ThroughputReliability = Readonly<{
  cv: number;
  iqrRatio: number;
  slopeNorm: number;
  label: ThroughputReliabilityLabel;
  samplesCount: number;
}>;

const PERCENTILE_KEYS: readonly PercentileKey[] = ["P50", "P70", "P90"];
const RELIABILITY_LABELS: readonly ThroughputReliabilityLabel[] = [
  "fiable",
  "incertain",
  "fragile",
  "non fiable",
];

function strictInteger(
  value: unknown,
  fieldName: string,
): number {
  if (typeof value !== "number" || !Number.isSafeInteger(value)) {
    throw new Error(`${fieldName} doit etre un entier strict.`);
  }
  return value;
}

function boundedInteger(
  value: unknown,
  fieldName: string,
  minimum: number,
  maximum: number,
): number {
  const resolved = strictInteger(value, fieldName);
  if (resolved < minimum || resolved > maximum) {
    throw new Error(`${fieldName} doit etre compris entre ${String(minimum)} et ${String(maximum)}.`);
  }
  return resolved;
}

function finiteMetric(value: unknown, fieldName: string): number {
  if (typeof value !== "number" || !Number.isFinite(value)) {
    throw new Error(`${fieldName} doit etre un nombre fini.`);
  }
  return value;
}

export function roundHalfUp(value: number, decimalPlaces = 4): number {
  const resolved = finiteMetric(value, "value");
  if (!Number.isSafeInteger(decimalPlaces) || decimalPlaces < 0) {
    throw new Error("decimalPlaces doit etre un entier strict >= 0.");
  }
  const [coefficient = "0", exponentText = "0"] = Math.abs(resolved)
    .toString()
    .toLowerCase()
    .split("e");
  const [integerPart = "0", fractionPart = ""] = coefficient.split(".");
  const digits = `${integerPart}${fractionPart}`;
  const decimalIndex = integerPart.length + Number(exponentText);
  const targetIndex = decimalIndex + decimalPlaces;
  const retainedDigits = targetIndex <= 0
    ? "0"
    : digits.slice(0, targetIndex).padEnd(targetIndex, "0");
  const firstDiscardedDigit = targetIndex < 0
    ? "0"
    : digits[targetIndex] ?? "0";
  let scaled = BigInt(retainedDigits);
  if (firstDiscardedDigit >= "5") scaled += 1n;
  const sign = resolved < 0 ? "-" : "";
  return Number(`${sign}${String(scaled)}e-${String(decimalPlaces)}`);
}

export function createSimulationSeed(value: unknown): SimulationSeed {
  return boundedInteger(
    value,
    "seed",
    SIMULATION_SEED_MIN,
    SIMULATION_SEED_MAX,
  ) as SimulationSeed;
}

export function createSimulationCount(value: unknown): SimulationCount {
  return boundedInteger(
    value,
    "n_sims",
    SIMULATION_N_SIMS_MIN,
    SIMULATION_N_SIMS_MAX,
  ) as SimulationCount;
}

export function createBacklogSize(value: unknown): BacklogSize {
  return boundedInteger(
    value,
    "backlog_size",
    SIMULATION_BACKLOG_SIZE_MIN,
    SIMULATION_BACKLOG_SIZE_MAX,
  ) as BacklogSize;
}

export function createSimulationHorizon(value: unknown): SimulationHorizon {
  return boundedInteger(
    value,
    "target_weeks",
    SIMULATION_TARGET_WEEKS_MIN,
    SIMULATION_HORIZON_WEEKS_MAX,
  ) as SimulationHorizon;
}

function validateRawThroughput(values: readonly unknown[]): readonly number[] {
  if (
    values.length < SIMULATION_THROUGHPUT_SAMPLES_MIN
    || values.length > SIMULATION_THROUGHPUT_SAMPLES_MAX
  ) {
    throw new Error(
      `throughput_samples doit contenir entre ${String(SIMULATION_THROUGHPUT_SAMPLES_MIN)} et ${String(SIMULATION_THROUGHPUT_SAMPLES_MAX)} valeurs.`,
    );
  }
  return values.map((value) => {
    const resolved = strictInteger(value, "throughput_samples");
    if (resolved < 0) {
      throw new Error("throughput_samples doit contenir uniquement des entiers >= 0.");
    }
    return resolved;
  });
}

export function createThroughputSamples(
  values: readonly unknown[],
  includeZeroWeeks: unknown,
): ThroughputSamples {
  if (typeof includeZeroWeeks !== "boolean") {
    throw new Error("include_zero_weeks doit etre un booleen strict.");
  }
  const rawValues = validateRawThroughput(values);
  const usableValues = includeZeroWeeks
    ? rawValues
    : rawValues.filter((value) => value > 0);
  if (usableValues.length < SIMULATION_THROUGHPUT_SAMPLES_MIN) {
    throw new Error(
      includeZeroWeeks
        ? "Historique insuffisant (moins de 6 semaines)."
        : "Historique insuffisant (moins de 6 semaines non nulles).",
    );
  }
  return Object.freeze({
    rawValues: Object.freeze([...rawValues]),
    usableValues: Object.freeze([...usableValues]),
    includeZeroWeeks,
  }) as ThroughputSamples;
}

function validatePercentileOrder(
  mode: SimulationMode,
  values: SimulationPercentiles,
): void {
  const present = PERCENTILE_KEYS.flatMap((key) => (
    values[key] === undefined ? [] : [values[key]]
  ));
  const isOrdered = present.slice(1).every((value, index) => (
    mode === "backlog_to_weeks"
      ? (present[index] ?? 0) <= value
      : (present[index] ?? 0) >= value
  ));
  if (!isOrdered) {
    const expected = mode === "backlog_to_weeks" ? "croissant" : "decroissant";
    throw new Error(`result_percentiles doit respecter l'ordre ${expected} du mode.`);
  }
}

export function createSimulationPercentiles(
  mode: SimulationMode,
  values: Readonly<Record<string, unknown>>,
): SimulationPercentiles {
  if (mode !== "backlog_to_weeks" && mode !== "weeks_to_items") {
    throw new Error("mode de simulation invalide.");
  }
  const unknownKeys = Object.keys(values).filter(
    (key) => !PERCENTILE_KEYS.includes(key as PercentileKey),
  );
  if (unknownKeys.length > 0) {
    throw new Error("result_percentiles accepte uniquement P50, P70 et P90.");
  }
  const resolved: Partial<Record<PercentileKey, number>> = {};
  PERCENTILE_KEYS.forEach((key) => {
    if (!(key in values)) return;
    const value = strictInteger(values[key], `result_percentiles.${key}`);
    if (value < 0) {
      throw new Error(`result_percentiles.${key} doit etre >= 0.`);
    }
    resolved[key] = value;
  });
  validatePercentileOrder(mode, resolved);
  return Object.freeze(resolved) as SimulationPercentiles;
}

export function riskScoreFromPercentiles(
  mode: SimulationMode,
  percentiles: SimulationPercentiles | null | undefined,
): number | undefined {
  if (!percentiles) return undefined;
  const validated = createSimulationPercentiles(mode, percentiles);
  const p50 = validated.P50;
  const p90 = validated.P90;
  if (p50 === undefined || p90 === undefined || p50 <= 0) return undefined;
  const numerator = mode === "weeks_to_items" ? p50 - p90 : p90 - p50;
  return roundHalfUp(Math.max(0, numerator / p50));
}

function categorizeReliability(
  cv: number,
  iqrRatio: number,
  slopeNorm: number,
  samplesCount: number,
  mean: number,
): ThroughputReliabilityLabel {
  if (samplesCount < 6 || mean <= 0 || cv >= 1.5 || slopeNorm <= -0.15) {
    return "non fiable";
  }
  if (cv >= 1 || iqrRatio >= 1 || Math.abs(slopeNorm) >= 0.1) {
    return "fragile";
  }
  if (cv >= 0.5 || iqrRatio >= 0.5 || Math.abs(slopeNorm) >= 0.05) {
    return "incertain";
  }
  return samplesCount < 8 ? "incertain" : "fiable";
}

type CalculatedReliability = {
  cv: unknown;
  iqrRatio: unknown;
  slopeNorm: unknown;
  samplesCount: unknown;
  mean: unknown;
};

type SerializedReliability = {
  cv: unknown;
  iqrRatio: unknown;
  slopeNorm: unknown;
  samplesCount: unknown;
  label: unknown;
};

export function createThroughputReliability(
  input: CalculatedReliability | SerializedReliability,
): ThroughputReliability {
  const cv = roundHalfUp(finiteMetric(input.cv, "cv"));
  const iqrRatio = roundHalfUp(finiteMetric(input.iqrRatio, "iqrRatio"));
  const slopeNorm = roundHalfUp(finiteMetric(input.slopeNorm, "slopeNorm"));
  if (cv < 0 || iqrRatio < 0) {
    throw new Error("cv et iqrRatio doivent etre >= 0.");
  }
  const samplesCount = strictInteger(input.samplesCount, "samplesCount");
  if (samplesCount < 0) {
    throw new Error("samplesCount doit etre >= 0.");
  }
  let label: ThroughputReliabilityLabel;
  if ("mean" in input) {
    label = categorizeReliability(
      cv,
      iqrRatio,
      slopeNorm,
      samplesCount,
      finiteMetric(input.mean, "mean"),
    );
  } else {
    if (
      typeof input.label !== "string"
      || !RELIABILITY_LABELS.includes(input.label as ThroughputReliabilityLabel)
    ) {
      throw new Error("label de fiabilite invalide.");
    }
    label = input.label as ThroughputReliabilityLabel;
  }
  return Object.freeze({
    cv,
    iqrRatio,
    slopeNorm,
    label,
    samplesCount,
  }) as ThroughputReliability;
}

export function createHistogram(
  buckets: readonly Readonly<{ x: unknown; count: unknown }>[],
  expectedMass: unknown,
): SimulationHistogram {
  const mass = strictInteger(expectedMass, "histogram.expectedMass");
  if (mass < 0) throw new Error("histogram.expectedMass doit etre >= 0.");
  const resolved = buckets.map((bucket) => {
    const x = strictInteger(bucket.x, "histogram.x");
    const count = strictInteger(bucket.count, "histogram.count");
    if (count <= 0) {
      throw new Error("histogram.count doit etre strictement positif.");
    }
    return Object.freeze({ x, count }) as HistogramBucket;
  });
  if (resolved.length > 100) {
    throw new Error("histogram doit contenir au plus 100 buckets.");
  }
  if (resolved.slice(1).some((bucket, index) => (resolved[index]?.x ?? bucket.x) >= bucket.x)) {
    throw new Error("histogram.x doit etre strictement croissant et sans doublon.");
  }
  if (resolved.reduce((sum, bucket) => sum + bucket.count, 0) !== mass) {
    throw new Error("histogram doit conserver sa masse totale.");
  }
  return Object.freeze(resolved) as SimulationHistogram;
}

export function createCompletionSummary(input: {
  completedCount: unknown;
  censoredCount: unknown;
  nSims: SimulationCount;
  horizonWeeks?: unknown;
  censoredRate?: unknown;
}): CompletionSummary {
  const nSims = createSimulationCount(input.nSims);
  const completedCount = strictInteger(input.completedCount, "completedCount");
  const censoredCount = strictInteger(input.censoredCount, "censoredCount");
  if (completedCount < 0 || censoredCount < 0) {
    throw new Error("Les comptes de completion doivent etre >= 0.");
  }
  if (completedCount + censoredCount !== nSims) {
    throw new Error("completedCount + censoredCount doit etre egal a nSims.");
  }
  const horizonWeeks = strictInteger(
    input.horizonWeeks ?? SIMULATION_HORIZON_WEEKS_MAX,
    "horizonWeeks",
  );
  if (horizonWeeks !== SIMULATION_HORIZON_WEEKS_MAX) {
    throw new Error("horizonWeeks doit etre egal a 521 pour le contrat 1.0.");
  }
  const censoredRate = roundHalfUp(censoredCount / nSims);
  if (
    input.censoredRate !== undefined
    && finiteMetric(input.censoredRate, "censoredRate") !== censoredRate
  ) {
    throw new Error("censoredRate doit etre derive des comptes.");
  }
  return Object.freeze({
    completedCount,
    censoredCount,
    censoredRate,
    horizonWeeks,
  }) as CompletionSummary;
}

File: backend/simulation_value_objects.py
from __future__ import annotations

import math
from collections.abc import Iterable, Iterator, Mapping, Sequence
from dataclasses import dataclass
from decimal import ROUND_HALF_UP, Decimal
from typing import Literal, TypeAlias, overload

from .risk_score import round_positive_ratio_half_up
from .simulation_limits import (
    SIMULATION_BACKLOG_SIZE_MAX,
    SIMULATION_BACKLOG_SIZE_MIN,
    SIMULATION_HORIZON_WEEKS_MAX,
    SIMULATION_N_SIMS_MAX,
    SIMULATION_N_SIMS_MIN,
    SIMULATION_SEED_MAX,
    SIMULATION_SEED_MIN,
    SIMULATION_TARGET_WEEKS_MIN,
    SIMULATION_THROUGHPUT_SAMPLES_MAX,
    SIMULATION_THROUGHPUT_SAMPLES_MIN,
)

SimulationMode: TypeAlias = Literal["backlog_to_weeks", "weeks_to_items"]
PercentileKey: TypeAlias = Literal["P50", "P70", "P90"]
ThroughputReliabilityLabel: TypeAlias = Literal["fiable", "incertain", "fragile", "non fiable"]

_PERCENTILE_KEYS: tuple[PercentileKey, ...] = ("P50", "P70", "P90")
_RELIABILITY_LABELS: frozenset[str] = frozenset({"fiable", "incertain", "fragile", "non fiable"})


class StatisticalValueError(ValueError):
    """Raised when a primitive cannot cross the statistical domain boundary."""


def _strict_integer(value: object, field_name: str) -> int:
    if type(value) is not int:
        raise StatisticalValueError(f"{field_name} doit etre un entier strict.")
    return value


def _bounded_integer(
    value: object,
    field_name: str,
    minimum: int,
    maximum: int,
) -> int:
    resolved = _strict_integer(value, field_name)
    if resolved < minimum or resolved > maximum:
        raise StatisticalValueError(f"{field_name} doit etre compris entre {minimum} et {maximum}.")
    return resolved


def _finite_metric(value: object, field_name: str) -> float:
    if isinstance(value, bool) or not isinstance(value, (int, float)):
        raise StatisticalValueError(f"{field_name} doit etre un nombre fini.")
    resolved = float(value)
    if not math.isfinite(resolved):
        raise StatisticalValueError(f"{field_name} doit etre un nombre fini.")
    return resolved


def round_half_up(value: float, decimal_places: int = 4) -> float:
    quantum = Decimal(1).scaleb(-decimal_places)
    return float(Decimal(str(value)).quantize(quantum, rounding=ROUND_HALF_UP))


@dataclass(frozen=True, slots=True)
class SimulationSeed:
    value: int

    def __post_init__(self) -> None:
        object.__setattr__(
            self,
            "value",
            _bounded_integer(
                self.value,
                "seed",
                SIMULATION_SEED_MIN,
                SIMULATION_SEED_MAX,
            ),
        )


@dataclass(frozen=True, slots=True)
class SimulationCount:
    value: int

    def __post_init__(self) -> None:
        object.__setattr__(
            self,
            "value",
            _bounded_integer(
                self.value,
                "n_sims",
                SIMULATION_N_SIMS_MIN,
                SIMULATION_N_SIMS_MAX,
            ),
        )


@dataclass(frozen=True, slots=True)
class BacklogSize:
    value: int

    def __post_init__(self) -> None:
        object.__setattr__(
            self,
            "value",
            _bounded_integer(
                self.value,
                "backlog_size",
                SIMULATION_BACKLOG_SIZE_MIN,
                SIMULATION_BACKLOG_SIZE_MAX,
            ),
        )


@dataclass(frozen=True, slots=True)
class SimulationHorizon:
    value: int

    def __post_init__(self) -> None:
        object.__setattr__(
            self,
            "value",
            _bounded_integer(
                self.value,
                "target_weeks",
                SIMULATION_TARGET_WEEKS_MIN,
                SIMULATION_HORIZON_WEEKS_MAX,
            ),
        )


@dataclass(frozen=True, slots=True, init=False)
class ThroughputSamples:
    raw_values: tuple[int, ...]
    usable_values: tuple[int, ...]
    include_zero_weeks: bool

    @classmethod
    def create(
        cls,
        values: Iterable[object],
        include_zero_weeks: bool,
    ) -> ThroughputSamples:
        if type(include_zero_weeks) is not bool:
            raise StatisticalValueError("include_zero_weeks doit etre un booleen strict.")
        try:
            raw_values = tuple(values)
        except TypeError as exc:
            raise StatisticalValueError("throughput_samples doit etre une collection.") from exc
        if not (
            SIMULATION_THROUGHPUT_SAMPLES_MIN
            <= len(raw_values)
            <= SIMULATION_THROUGHPUT_SAMPLES_MAX
        ):
            raise StatisticalValueError(
                "throughput_samples doit contenir entre "
                f"{SIMULATION_THROUGHPUT_SAMPLES_MIN} et "
                f"{SIMULATION_THROUGHPUT_SAMPLES_MAX} valeurs."
            )

        validated = tuple(_strict_integer(value, "throughput_samples") for value in raw_values)
        if any(value < 0 for value in validated):
            raise StatisticalValueError(
                "throughput_samples doit contenir uniquement des entiers >= 0."
            )
        usable = (
            validated if include_zero_weeks else tuple(value for value in validated if value > 0)
        )
        if len(usable) < SIMULATION_THROUGHPUT_SAMPLES_MIN:
            detail = (
                "Historique insuffisant (moins de 6 semaines)."
                if include_zero_weeks
                else "Historique insuffisant (moins de 6 semaines non nulles)."
            )
            raise StatisticalValueError(detail)
        instance = object.__new__(cls)
        object.__setattr__(instance, "raw_values", validated)
        object.__setattr__(instance, "usable_values", usable)
        object.__setattr__(instance, "include_zero_weeks", include_zero_weeks)
        return instance


@dataclass(frozen=True, slots=True, init=False)
class SimulationPercentiles(Mapping[PercentileKey, int]):
    mode: SimulationMode
    _items: tuple[tuple[PercentileKey, int], ...]

    @classmethod
    def create(
        cls,
        mode: SimulationMode,
        values: Mapping[str, object],
    ) -> SimulationPercentiles:
        if mode not in ("backlog_to_weeks", "weeks_to_items"):
            raise StatisticalValueError("mode de simulation invalide.")
        unknown_keys = set(values) - set(_PERCENTILE_KEYS)
        if unknown_keys:
            raise StatisticalValueError("result_percentiles accepte uniquement P50, P70 et P90.")
        normalized: dict[PercentileKey, int] = {}
        for key in _PERCENTILE_KEYS:
            if key not in values:
                continue
            value = _strict_integer(values[key], f"result_percentiles.{key}")
            if value < 0:
                raise StatisticalValueError(f"result_percentiles.{key} doit etre >= 0.")
            normalized[key] = value
        cls._validate_order(mode, normalized)
        instance = object.__new__(cls)
        object.__setattr__(instance, "mode", mode)
        object.__setattr__(instance, "_items", tuple(normalized.items()))
        return instance

    @staticmethod
    def _validate_order(
        mode: SimulationMode,
        values: Mapping[PercentileKey, int],
    ) -> None:
        ordered_values = [values[key] for key in _PERCENTILE_KEYS if key in values]
        pairs = zip(ordered_values, ordered_values[1:])
        valid = (
            all(left <= right for left, right in pairs)
            if mode == "backlog_to_weeks"
            else all(left >= right for left, right in pairs)
        )
        if not valid:
            expected = "croissant" if mode == "backlog_to_weeks" else "decroissant"
            raise StatisticalValueError(
                f"result_percentiles doit respecter l'ordre {expected} du mode."
            )

    @overload
    def __getitem__(self, key: PercentileKey) -> int: ...

    @overload
    def __getitem__(self, key: str) -> int: ...

    def __getitem__(self, key: str) -> int:
        return dict(self._items)[key]

    def __iter__(self) -> Iterator[PercentileKey]:
        return (key for key, _value in self._items)

    def __len__(self) -> int:
        return len(self._items)

    def __eq__(self, other: object) -> bool:
        if isinstance(other, Mapping):
            return dict(self._items) == dict(other)
        return False

    def to_dict(self) -> dict[str, int]:
        return dict(self._items)

    @property
    def risk_score(self) -> float | None:
        p50 = self.get("P50")
        p90 = self.get("P90")
        if p50 is None or p90 is None or p50 <= 0:
            return None
        numerator = p50 - p90 if self.mode == "weeks_to_items" else p90 - p50
        return round_positive_ratio_half_up(max(0, numerator), p50)


def _categorize_reliability(
    cv: float,
    iqr_ratio: float,
    slope_norm: float,
    samples_count: int,
    mean: float,
) -> ThroughputReliabilityLabel:
    if samples_count < 6 or mean <= 0 or cv >= 1.5 or slope_norm <= -0.15:
        return "non fiable"
    if cv >= 1 or iqr_ratio >= 1 or abs(slope_norm) >= 0.1:
        return "fragile"
    if cv >= 0.5 or iqr_ratio >= 0.5 or abs(slope_norm) >= 0.05:
        return "incertain"
    return "incertain" if samples_count < 8 else "fiable"


@dataclass(frozen=True, slots=True, init=False)
class ThroughputReliability:
    cv: float
    iqr_ratio: float
    slope_norm: float
    label: ThroughputReliabilityLabel
    samples_count: int

    @classmethod
    def create(
        cls,
        *,
        cv: object,
        iqr_ratio: object,
        slope_norm: object,
        samples_count: object,
        mean: object | None = None,
        label: object | None = None,
    ) -> ThroughputReliability:
        normalized_cv = round_half_up(_finite_metric(cv, "cv"))
        normalized_iqr = round_half_up(_finite_metric(iqr_ratio, "iqr_ratio"))
        normalized_slope = round_half_up(_finite_metric(slope_norm, "slope_norm"))
        if normalized_cv < 0 or normalized_iqr < 0:
            raise StatisticalValueError("cv et iqr_ratio doivent etre >= 0.")
        resolved_count = _strict_integer(samples_count, "samples_count")
        if resolved_count < 0:
            raise StatisticalValueError("samples_count doit etre >= 0.")

        if mean is not None:
            resolved_mean = _finite_metric(mean, "mean")
            resolved_label = _categorize_reliability(
                normalized_cv,
                normalized_iqr,
                normalized_slope,
                resolved_count,
                resolved_mean,
            )
        else:
            if not isinstance(label, str) or label not in _RELIABILITY_LABELS:
                raise StatisticalValueError("label de fiabilite invalide.")
            resolved_label = label

        instance = object.__new__(cls)
        object.__setattr__(instance, "cv", normalized_cv)
        object.__setattr__(instance, "iqr_ratio", normalized_iqr)
        object.__setattr__(instance, "slope_norm", normalized_slope)
        object.__setattr__(instance, "label", resolved_label)
        object.__setattr__(instance, "samples_count", resolved_count)
        return instance


@dataclass(frozen=True, slots=True, init=False)
class HistogramBucket:
    x: int
    count: int

    @classmethod
    def create(cls, x: object, count: object) -> HistogramBucket:
        resolved_x = _strict_integer(x, "histogram.x")
        resolved_count = _strict_integer(count, "histogram.count")
        if resolved_count <= 0:
            raise StatisticalValueError("histogram.count doit etre strictement positif.")
        instance = object.__new__(cls)
        object.__setattr__(instance, "x", resolved_x)
        object.__setattr__(instance, "count", resolved_count)
        return instance


@dataclass(frozen=True, slots=True, init=False)
class Histogram(Sequence[HistogramBucket]):
    buckets: tuple[HistogramBucket, ...]

    @classmethod
    def create(
        cls,
        buckets: Iterable[Mapping[str, object] | HistogramBucket],
        *,
        expected_mass: int,
    ) -> Histogram:
        mass = _strict_integer(expected_mass, "histogram.expected_mass")
        if mass < 0:
            raise StatisticalValueError("histogram.expected_mass doit etre >= 0.")
        resolved = tuple(
            bucket
            if isinstance(bucket, HistogramBucket)
            else HistogramBucket.create(bucket.get("x"), bucket.get("count"))
            for bucket in buckets
        )
        if len(resolved) > 100:
            raise StatisticalValueError("histogram doit contenir au plus 100 buckets.")
        if any(left.x >= right.x for left, right in zip(resolved, resolved[1:])):
            raise StatisticalValueError(
                "histogram.x doit etre strictement croissant et sans doublon."
            )
        if sum(bucket.count for bucket in resolved) != mass:
            raise StatisticalValueError("histogram doit conserver sa masse totale.")
        instance = object.__new__(cls)
        object.__setattr__(instance, "buckets", resolved)
        return instance

    @overload
    def __getitem__(self, index: int) -> HistogramBucket: ...

    @overload
    def __getitem__(self, index: slice) -> tuple[HistogramBucket, ...]: ...

    def __getitem__(self, index: int | slice) -> HistogramBucket | tuple[HistogramBucket, ...]:
        return self.buckets[index]

    def __len__(self) -> int:
        return len(self.buckets)


@dataclass(frozen=True, slots=True, init=False)
class CompletionSummary:
    completed_count: int
    censored_count: int
    censored_rate: float
    horizon_weeks: int

    @classmethod
    def create(
        cls,
        *,
        completed_count: object,
        censored_count: object,
        n_sims: SimulationCount,
        horizon_weeks: object = SIMULATION_HORIZON_WEEKS_MAX,
    ) -> CompletionSummary:
        if not isinstance(n_sims, SimulationCount):
            raise StatisticalValueError("n_sims doit etre un Value Object.")
        completed = _strict_integer(completed_count, "completed_count")
        censored = _strict_integer(censored_count, "censored_count")
        if completed < 0 or censored < 0:
            raise StatisticalValueError("Les comptes de completion doivent etre >= 0.")
        if completed + censored != n_sims.value:
            raise StatisticalValueError("completed_count + censored_count doit etre egal a n_sims.")
        horizon = _strict_integer(horizon_weeks, "horizon_weeks")
        if horizon != SIMULATION_HORIZON_WEEKS_MAX:
            raise StatisticalValueError("horizon_weeks doit etre egal a 521 pour le contrat 1.0.")
        rate = round_half_up(censored / n_sims.value)
        instance = object.__new__(cls)
        object.__setattr__(instance, "completed_count", completed)
        object.__setattr__(instance, "censored_count", censored)
        object.__setattr__(instance, "censored_rate", rate)
        object.__setattr__(instance, "horizon_weeks", horizon)
        return instance
